client/drain.go

```diff
@@ -0,0 +1,88 @@
+package client
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+func (c *Client) DrainSelf() error {
+	drainSpec := c.GetConfig().Drain
+	if drainSpec == nil {
+		return nil
+	}
+
+	logger := c.logger.With("drain")
+
+	now := time.Now()
+	drainReq := &structs.NodeUpdateDrainRequest{
+		NodeID: c.NodeID(),
+		DrainStrategy: &structs.DrainStrategy{
+			DrainSpec: structs.DrainSpec{
+				Deadline:         drainSpec.Deadline,
+				IgnoreSystemJobs: drainSpec.IgnoreSystemJobs,
+			},
+			StartedAt: now,
+		},
+		MarkEligible: false,
+		Meta:         map[string]string{"message": "shutting down"},
+		WriteRequest: structs.WriteRequest{
+			Region: c.Region(), AuthToken: c.secretNodeID()},
+	}
+	if drainSpec.Deadline > 0 {
+		drainReq.DrainStrategy.ForceDeadline = now.Add(drainSpec.Deadline)
+	}
+
+	var drainResp *structs.NodeDrainUpdateResponse
+	err := c.RPC("Node.UpdateDrain", drainReq, drainResp)
+	if err != nil {
+		logger.Error("failed to drain self on shutdown", "error", err)
+		return err
+	}
+
+	// note: letting this run forever seems wrong but init system (ex systemd)
+	// will almost always force kill the client eventually
+	ctx := context.Background()
```

Comment:
Of course, we could always cap this to something really long like 24h, so long as we document that behavior in the config docs.

Reply:
Switched to 1hr.